### ISSUE TITLE
Principal Propagation Deployment Issues

### ISSUE THREAD

**Issue by SteveButch** [IBM-i2/analyze-connect#5]
I am struggling to get the example PrincipalPropagation going in my environment.  I have followed through the instructions, can compile the project and deploy it but when I run my connectors I do not get anything added to the request header.  I have been looking through the log files and don't see any errors being thrown so I am but stuck on next steps to first make sure that I have properly deployed the project and then to figure out why the headers are not being added.

Any ideas on where I should best start?

**Comment by TonyJon** [comment]
Hi Steve

I haven't tried doing this myself but will see if I can get information from some of the devs  for how to best proceed.

In the mean time...

Looking at what the example says, I would initially check that the created JAR has actually been put here:

toolkit/configuration/fragments/opal-services/WEB-INF/lib/ directory of the i2 Analyze toolkit specified in the <toolkitLocation> element

and that after you modified the 

toolkit/configuration/fragments/opal-services/WEB-INF/classes/DiscoServerSettingsCommon.properties file 

to have this at the end:

ConnectorRequestModifier=com.i2group.example.PrincipalPropagation

You then redeployed and restarted Liberty.

Cheers

**Comment by TonyJon** [comment]
Hi Again Steve
One of our developers shared this with me:

_An easy way to check whether it's the config or the implementation is to set ConnectorRequestModifier=not.a.real.Class  and redeploy and restart  Liberty then the server should blow up at startup / when making a connector request._

If it doesn't then either the config isn't actually being used by the application. (Or the file you have set it in isn't)._

What version  of i2 Connect and  i2Analyze/EIA are you using?

**Comment by TonyJon** [comment]
Hi Steve 

I grabbed the Principal Propagation project and  followed the documentation to create/deploy and reference the default PrincipalPropagation.jar  so that I should see the extra header named i2ExtensionHeader in responses from the i2 Connect gateway when using my example connector.

But I was not looking for it in the correct way so did not see any difference in the header content.

I just looked at the browser communication going from ANBP when I ran a search and didn't see it, but that is not where it is added. (Did you do this as well?)

Basically  the same  request from ANBP to the server  is  made by the server to the connector and that is when we add the extra header information.

So to see this I/we would have to look at what the connector is receiving not what the ANBP end is sending.

Apologies if that is what you are already doing, and if that is the case hopefully my earlier comments may have helped you get to the bottom of your issue.

Cheers

**Comment by SteveButch** [comment]
Working through the suggestions....

I have been modifying the DiscoServerSettingsCommon.properties in \i2analyze\toolkit\configuration\opal-services\WEB-INF\classes\ to add the ConnectorRequestModifier=com.i2group.example.PrincipalPropagation line.

I must have read and reread this 100 times, ConnectorRequestModifier that is the correct entry right?

I stopped and deployed Liberty and noted that the file was also updated in \i2analyze\deploy\wlp\usr\servers\opal-server\apps\opal-services-is.war\WEB-INF\classes\

Starting Liberty but with a purposely incorrect name did not seem to generate any errors either on the EIA server or the connector, the server comes up clean and the connector worked as it did before any changes to the properties file.
We are on i2Analyze 4.3.1 (not sure where to look for the i2 Connect version)

We are using ANBP so I have not been looking at the communication between ANBP and the server.  I am looking at the request and response headers inside my connector, so I guess the communications between the EIA server and my NodeJS endpoint.  In my endpoint (we are using the Node example) for the request I only see the basic entries of content-type, accept-language, agent, length etc.  I do not see the i2ExtensionHeader.

**Comment by TonyJon** [comment]
You can check this out for yourself by adding some logging into your connector. (Thank you to the Developer that suggested this).

I am using the example node.js one that we ship so added  this to its app.js:

```
function logrequest(req) {
	console.log('**************');
	console.log(JSON.stringify(req.url));
	console.log(JSON.stringify(req.headers));
}
```
And then added  a call to that in my app.post for  the example search validate endpoint:

```
app.post("/exampleSearch/validate", (req, res) => {
  // If there were validation issues with the payload and/or the connector,
  // the error message to display to the user would be set here
  const errorMessage = undefined;

  const response = errorMessage ? { errorMessage } : {};
  logrequest(req)
  res.send(JSON.stringify(response));
});
```

When I make an External Search from ANBP my connector console now shows me:

```
"/exampleSearch/validate"
{"content-type":"application/json","accept-language":"en-GB","i2extensionheader":"Jenny","content-length":"93","host":"localhost:3700","connection":"Keep-Alive","user-agent":"Apache-HttpClient/4.5.12 (Java/1.8.0_232)","accept-encoding":"gzip,deflate"}
```

**Comment by TonyJon** [comment]
Hi Steve

Sorry your reply didn't show until after I made that last post. it is past my end of Day here now as I start at  7am UK time so I will be heading out now.

I will check with the devs but it could be that 4.3.1 was before  we implemented the ability to use this functionality.

**Comment by SteveButch** [comment]
While frustrating that I was not working since we are on the wrong version it is nice to know I can at least follow directions.  Thank you all for the info, I will close this out and see about getting our environment upgraded.

Be well,
Steve

**Comment by TonyJon** [comment]
Thank you Steve, and yes you were spot on with what you did.

If you can delay your environment upgrade by a few weeks, we are putting the finishing touches on our latest set of 2021 products at the moment and moving to those, when they come out, will give you an even greater set of capabilities that you can take advantage of.

Cheers
